core/src/replay_stage.rs

```diff
@@ -395,6 +403,11 @@ impl ReplayStage {
                     );
                     replay_active_banks_time.stop();
 
+                    // persist cost table to blockstore
+                    let mut persist_cost_table_time = Measure::start("persist_cost_table_time");
+                    Self::persist_cost_table(&blockstore, &cost_model);
```

Comment:
Is there a need to write the cost table if there were no updates? It's not entirely uncommon for some iterations of this loop to not replay anything if there have been no new connected blocks/shreds that have arrived.

Reply:
Valid point. Will included in next iteration

Reply:
Piggy-backing off of @carllin's comment, is it the case that the entire cost table would need to re-written, or could it be the case that the cost table may be partially updated (some entries updated / some not)?

Assuming the table could be partially updated, we'd ideally only write the programs that were updated . It seems like you'd have to track the list of updates, or update the structure to include a dirty bit, something like this:
```
        for (key, cost_entry) in cost_table.iter() {
            if cost_entry.dirty { 
                blockstore
                    .write_program_cost(key, cost_entry.value)
                    .expect("persist program costs to blockstore");
            }
        }
```
Not sure the overhead would justify any potential gains since the table is relatively small, or if a "partial update" is even a valid scenario, but just thinking out loud

Reply:
Totally valid input. As described early, the cost table has fixed capacity, the old/infrequent program will be pushed out of table to make room for more recent updates. This is to prevent cost table eats up all memories. 

For example, if cost table is initialized with capacity=3, at one point it has content: `[A, B, C]`. Later `B` is updated and `D` is added, which pushes out `A`, the table would look like this: `[B', C, D]`. When update blockstore, it needs three ops: remove A, update B, insert D. 

For the sake of simplicity, I am taking approach of truncating column family before writing new table (eg table has been updated somehow) to blockstore. With the small size of table, plus moving all cost_model related operations off replay main thread to its own thread, I'd think this is acceptable. But I am not confident if this `erase, write column family` operation will not harm blockstore performance. Thoughts?

Reply:
in the end, I added `delete_cf` to `Rocks` and `LedgerColumn` so `replay_stage` can delete those programs no longer in memory( cost table).